### ISSUE TITLE
Support for Async Performers in Effect Library

### ISSUE THREAD

**Issue by giuliohome** [python-effect/effect#86]
Hello,

I’m currently implementing an async performer, starting from a synchronous version. In the synchronous setup, I have:

```python
result = sync_perform(mock_dispatcher, main_sequence(logger))
```

This approach uses a pure logic sequence for managing effects in `main_sequence`, which can be found [here](https://github.com/giuliohome-org/effectful-api/blob/08059c1cfc788ff1b4a0a7206b4f4d510133804f/main.py#L14C5-L14C66).

For the async version, I have created a custom `async_perform` function to handle asynchronous performers, defined  [here](https://github.com/giuliohome-org/effectful-api/blob/2bca82b937d18f6e2fdecdc9a4e04907a6876df0/async_monad.py#L6-L43).

### Usage
In the async main, the call looks like this:
```python
result = await async_perform(mock_dispatcher, main_sequence(logger))
```
Here’s an example of an async performer:
```python
# ...
    async def mock_create_request_performer(dispatcher, intent):
        logger.debug(f"Mock: Creating request... {intent.payload} with token {token}")
        await asyncio.sleep(1)
# ...
    return TypeDispatcher({
        CreateRequest: mock_create_request_performer,
# ...
```

Look at my [feat/async](https://github.com/giuliohome-org/effectful-api/tree/feat/async) branch for the whole sample.

Would this be considered the recommended approach for handling async performers, or are there any improvements or best practices that you’d suggest?

Thank you!

**Comment by radix** [comment]
@giuliohome just so you know, this library is effectively abandoned, since I haven't touched it or even really written any Python in several years.

You should look at the way [txeffect](https://github.com/python-effect/txeffect) implements its [txeffect.perform method](https://github.com/python-effect/txeffect/blob/master/txeffect/__init__.py#L88). Ideally, this would be ported to python's built-in async system and included directly in the main `effect` library.

Your implementation seems to be reimplementing all of [effect.perform](https://github.com/python-effect/effect/blob/master/effect/_base.py#L80) (which is separate from effect.sync_perform), but you should just be able to invoke it instead of reimplementing it. There are some situations that effect.perform will automatically solve for you like avoiding extraneous stack usage by way of a [trampoline](https://github.com/python-effect/effect/blob/master/effect/_base.py#L155). 

Another intentional design is that performers get decorated either in a [@sync_performer](https://github.com/python-effect/effect/blob/master/effect/_sync.py#L36) or a [@deferred_performer](https://github.com/python-effect/txeffect/blob/master/txeffect/__init__.py#L46), which again I would port to a `@async_performer` decorator that does the same thing. Maybe this could be rethought so that the decorator isn't needed, but that's the idiom that effect goes with.

**Comment by giuliohome** [comment]
Yes, thank you very much. I understand what you suggest, you're right, indeed I tried to do that initially, but I failed somewhere and resorted to the full implementation, because I was in a hurry. I will try again soon, thank you for the kind and prompt reply.

**Comment by giuliohome** [comment]
I've reviewed it again, but unfortunately, it doesn't seem possible to do it that way. In a word, the core issue here is the async '_color_' problem, which fundamentally sets it apart from a pure callback

**Comment by giuliohome** [comment]
**If both the trampoline and the bouncer are made async,** then this async performer will function as expected

```py
def async_performer(f):
    @wraps(f)
    async def async_wrapper(*args, **kwargs):
        box = args[-1]
        pass_args = args[:-1]
        try:
            print(f"Performing effect: {f} in async_performer!!!")
            result = await f(*pass_args, **kwargs)
            await box.succeed(result)
        except Exception as e:
            box.fail(e)

    return async_wrapper
```


I will fork this library to show the async version.
